org.eclipse.xtext/src/org/eclipse/xtext/formatting2/AbstractJavaFormatter.java

```diff
@@ -0,0 +1,162 @@
+import org.eclipse.xtext.formatting2.regionaccess.IHiddenRegion;
+import org.eclipse.xtext.formatting2.regionaccess.ISemanticRegion;
+import org.eclipse.xtext.formatting2.regionaccess.ISemanticRegionFinder;
+import org.eclipse.xtext.formatting2.regionaccess.ISemanticRegionsFinder;
+import org.eclipse.xtext.formatting2.regionaccess.ITextRegionExtensions;
+import org.eclipse.xtext.resource.XtextResource;
+import org.eclipse.xtext.util.PolymorphicDispatcher;
+
+import com.google.common.annotations.Beta;
+
+/**
+ * <p>
+ * This is an abstract base class for language-specific formatters, intended to be extended with a java class.
+ * </p>
+ * <p>
+ * It is working very much like {@link AbstractFormatter2} but does support additional methods to simplify java
+ * formatter code.
+ * </p>
+ * <p>
+ * the "_format" methods are called by reflection. The name is important as well as there are exactly two arguments
+ * whereas the first one is of the element type you like to format and the second one of type
+ * {@link IFormattableDocument}. The methods should be protected.
+ * </p>
+ * <p>
+ * Example code for a java formatter:
+ * 
+ * <pre>
+ * import static org.xtext.example.mydsl.myDsl.MyDslPackage.Literals.*;
+ * 
+ * public class MyDslFormatter2 extends AbstractJavaFormatter {
+ * 	protected void _format(Model model, IFormattableDocument doc) {
+ * 		for (Parent parent : model.getParents())
+ * 			doc.format(parent);
+ * 	}
+ * 
+ * 	protected void _format(Parent parent, IFormattableDocument doc) {
+ * 		doc.prepend(regionFor(parent).keyword("parent"), it -> it.noSpace());
+ * 		doc.append(regionFor(parent).keyword("parent"), it -> it.oneSpace());
+ * 		doc.append(regionFor(parent).feature(PARENT__NAME), it -> it.oneSpace());
+ * 		doc.prepend(regionFor(parent).keyword("{"), it -> it.oneSpace());
+ * 		doc.append(regionFor(parent).keyword("{"), it -> it.newLine());
+ * 		doc.interior(regionFor(parent).keyword("{"), regionFor(parent).keyword("}"), it -> it.indent());
+ * 		doc.append(regionFor(parent).keyword("}"), it -> it.setNewLines(1, 1, 2));
+ * 		for (Child child : parent.getChildren())
+ * 			doc.format(child);
+ * 	}
+ * 
+ * 	protected void _format(Child child, IFormattableDocument doc) {
+ * 		doc.append(regionFor(child).keyword("child"), it -> it.oneSpace());
+ * 		doc.append(regionFor(child).feature(CHILD__NAME), it -> it.newLine());
+ * 	}
+ * }
+ * </pre>
+ * </p>
+ * <p>
+ * This is suitable for the following grammar:
+ * 
+ * <pre>
+ * grammar org.xtext.example.mydsl.MyDsl with org.eclipse.xtext.common.Terminals
+ * 
+ * generate myDsl "http://www.xtext.org/example/mydsl/MyDsl"
+ * 
+ * Model: parents+=Parent*;
+ * Parent: 'parent' name=ID '!' ('{' children+=Child* '}')?;
+ * Child: 'child' name=ID;
+ * </pre>
+ * </p>
+ */
+@Beta
+public abstract class AbstractJavaFormatter extends AbstractFormatter2 implements ITextRegionExtensions {
+
+	private PolymorphicDispatcher<Void> dispatcher = PolymorphicDispatcher
```

Comment:
Should be fina, protected and instantiated via a `createPolymorhicDispatcher` method that can be overridden if someone prefers a different naming scheme, annotations, more arguments (settings or similar) to be passed around.

Reply:
Ok, will do ... by the way, I named the methods "_format", not "format" ... not sure about if it is necessary. What do you think?